src/par/src/TPPartitioner.cpp

```diff
@@ -380,34 +367,109 @@ void TPpartitioner::OptimalPartCplexWarmStart(
+      for (auto eptr_id = hgraph->eptr_[e]; eptr_id < hgraph->eptr_[e+1];
+           eptr_id++) {
+        vertex_mask.insert(hgraph->eind_[eptr_id]);
+      }     
+    } else {
+      break; // the set has been sorted
+    } 
+  }
+  // To make sure that there is always only one optimal solution,
+  // while preserving the best structure
+  // the min_diff_score will be added to the vertices within one block
+  const float min_diff_score = 1.0f / (vertex_mask.size() * vertex_mask.size() * num_parts_); 
+  logger_->report("min_diff_score : {}", min_diff_score);
+  // generate reduced hypergraph and store the mapping relationship
+  std::vector<int> old_vertex_vec(hgraph->num_vertices_, -1); // map old vertex id to new vertex id
+  std::vector<int> new_vertex_vec(vertex_mask.size(), -1); // map new vertex id to old vertex id
+  std::vector<std::set<int> > new_hyperedges; // new hyperedges
+  const int num_vertices = vertex_mask.size();
+  const int num_hyperedges = edge_mask.size();
+  // create new vertices
+  int vertex_id = 0;
+  for (auto& old_vertex : vertex_mask) {
+    old_vertex_vec[old_vertex] = vertex_id;
+    new_vertex_vec[vertex_id++] = old_vertex;
+  }
+  // create new hyperedges
+  for (auto& e : edge_mask) {
+    std::set<int> hyperedge;
+    for (auto eptr_id = hgraph->eptr_[e]; eptr_id < hgraph->eptr_[e+1];
+         eptr_id++) {
+      hyperedge.insert(old_vertex_vec[hgraph->eind_[eptr_id]]);
+    }
+    // Note that here we do not need to detect whether the hyperegde is a single-vertex hyperedge
+    // Because we just rename the vertices, thus the dimensionality of each hyperedge will not change
+    new_hyperedges.push_back(hyperedge);
+  }
+  // print basic information
+  logger_->report("After Hyperedge Reduction :  num_vertices = {}, num_hyperedges = {}",
+                   num_vertices,  num_hyperedges);
+
+  // From here, we are going to create CP-SAT model for the reduced hypergraph
+  matrix<int> x(num_parts_, std::vector<int>(num_vertices, 0));
+  matrix<int> y(num_parts_, std::vector<int>(num_hyperedges, 0));
+  // set hints for vertices and hyperedges
+  // set vertex v to its partition solution
+  for (auto v = 0; v < num_vertices; v++) {
+    const int part_id = solution[new_vertex_vec[v]];
+    for (auto i = 0; i < num_parts_; i++) {
+      x[i][v] = (i == part_id) ? 1 : 0; 
+    }
+  }  
+  // set hyperedge e to its partition solution
+  // if hyperedge e is fully with partition i, set y[i][e] = 1
+  // otherwise y[i][e] = 0
+  for (auto e = 0; e < num_hyperedges; e++) {
+    std::set<int> unique_partitions;
+    for (auto& v : new_hyperedges[e]) {
+      unique_partitions.insert(solution[new_vertex_vec[v]]);
+    }
+    if (unique_partitions.size() == 1) {
+      const int part_id = *unique_partitions.begin();
+      for (int i = 0; i < num_parts_; i++) {
+        y[i][e] = (i == part_id) ? 1 : 0;
+      }            
+    } else {
+      for (int i = 0; i < num_parts_; i++) {
+        y[i][e] = 0;
+      }
```

Comment:
y was already initialized to 0 so this seems to be a no-op

Reply:
This line was written to make the algorithm more clear.  (If the hyperedge spans multiple partitions,  we need to set y to 0)